decidim/decidim
# Add the key ceremony

#### :tophat: What? Why?
This PR allows a trustee to generate election keys for an election he/she is a trustee for. During the key ceremony process, the trustee has to download their backup keys. When **all** trustees completed that step, the process continues. If one trustee leaves the page after generating the backup keys, he/she has to restore the keys to continue the process. 

#### :pushpin: Related Issues
- Fixes #6363 

#### Testing
_GENERATING ELECTION KEYS_
1. As an admin, create two new trustees (`admin@example.org` and `user@example.org`) and set up an election with these two trustees.
2. Go to the trustee zone for trustee 1 (Admin). You should find the election and the link **Action required**. Following this link, you enter the key ceremony page. Click the **Start** button and download the**Backup keys**.
3. Go to the trustee zone for the 2nd trustee and repeat the steps above. You should see a success notification with the new election bulletin board status: `ready`. 
4. Click the **Back** button. 
5. You should see an updated list of elections, showing the new `bb_status`. The link **Action required** is no longer available. Instead, you should see **No**.

_GENERATING ELECTION KEYS WITH RESTORING BACKUP KEYS_
1. As an admin, create two new trustees (`admin@example.org` and `user@example.org`) and set up an election with these two trustees.
2. Go to the trustee zone for trustee 1 (Admin). You should find the election and the link **Action required**. Following this link, you enter the key ceremony page. Click the **Start** button and download the**Backup keys**.
3. Leave the page, e.g. go back to the trustee zone election overview (previous page).
4. Click **Action required** again.
5. Click *Start**.
6. Upload your **Backup keys**.
7. Go to the trustee zone for the 2nd trustee and repeat the steps above. You should see a success notification with the new election bulletin board status: `ready`. 
8. Click the **Back** button. 
9. You should see an updated list of elections, showing the new `bb_status`. The link **Action required** is no longer available. Instead, you should see **No**.

_CHECKING GRAPHQL_
As soon as the `key ceremony` is done, you can open GraphiQl (Bulletin board) and run the following query, where `uniqueId` is equal to your Decidim Authority (parameterized), followed by a dot and the id of the `election`. 

```
query {
    election(uniqueId: "decidim-test-authority.13") {
    logEntries {
      messageId
      client {
        name
      }
      chainedHash
    }
  }
}
```
#### :clipboard: Checklist
:rotating_light: Please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.


### :camera: Screenshots
*Please add screenshots of the changes you're proposing*
**Trustee zone with Elections**
<img width="1400" alt="Trustee Zone Election Overview" src="https://user-images.githubusercontent.com/19774772/102875054-921ffe80-4443-11eb-98f0-40ad4e1f3aee.png">

**Key Ceremony - Generate election keys**
<img width="1377" alt="Trustee Zone Generate Election Keys" src="https://user-images.githubusercontent.com/19774772/102875912-bcbe8700-4444-11eb-9954-bff6c45fb670.png">

**Key Ceremony - Backup election keys**
<img width="1359" alt="Trustee Zone Download Backup Keys" src="https://user-images.githubusercontent.com/19774772/102875933-c6e08580-4444-11eb-8ac6-fddb40cf8718.png">

**Key Ceremony - Restore election keys**
<img width="1727" alt="Trustee Zone Restore Election Keys" src="https://user-images.githubusercontent.com/19774772/102875948-cc3dd000-4444-11eb-86f5-e98b9cab45b5.png">

**Key Ceremony - Success**
<img width="1389" alt="Trustee Zone Key Ceremony Success" src="https://user-images.githubusercontent.com/19774772/102875961-d3fd7480-4444-11eb-964d-5b1168d32dbb.png">

:hearts: Thank you!
